testutil/validatormock/component.go

```diff
@@ -0,0 +1,452 @@
+			return err
+		}
+	case core.DutySyncMessage:
+		if syncComm == nil {
+			return errors.New("syncomm is nil", z.Str("duty", duty.String()))
+		}
+
+		if err = syncComm.Message(ctx, eth2Slot); err != nil { // Rename to sync.Comm.SyncMessage
+			return err
+		}
+	case core.DutySyncContribution:
+		if syncComm == nil {
+			return errors.New("syncomm is nil", z.Str("duty", duty.String()))
+		}
+
+		if _, err = syncComm.Aggregate(ctx, eth2Slot); err != nil { // Rename to sync.Comm.AggregateSyncContribution
+			return err
+		}
+	default:
+		return errors.New("unexpected duty", z.Str("duty", duty.String()))
+	}
+
+	return nil
+}
+
+func (m *Component) startAttesters(epoch metaEpoch) error {
+	for _, slot := range epoch.Slots() {
+		eth2Cl, err := m.eth2ClProvider()
+		if err != nil {
+			return err
+		}
+
+		attester := NewSlotAttester(eth2Cl, eth2p0.Slot(slot.Slot), m.signFunc, m.pubkeys)
+		m.setAttester(attester)
+	}
+
+	return nil
+}
+
+func (m *Component) startSyncCommMembers(ctx context.Context, epoch metaEpoch) error {
+	eth2Cl, err := m.eth2ClProvider()
+	if err != nil {
+		return err
+	}
+
+	syncCommMem := NewSyncCommMember(eth2Cl, eth2p0.Epoch(epoch.Epoch), m.signFunc, m.pubkeys)
+	if err = syncCommMem.PrepareEpoch(ctx); err != nil {
+		return err
+	}
+
+	m.setSyncCommMember(syncCommMem)
+
+	return nil
+}
+
+func (m *Component) deleteAttesters(epoch metaEpoch) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	for _, slot := range epoch.Slots() {
+		delete(m.attestersBySlot, slot.Slot)
+	}
+}
+
+func (m *Component) deleteSyncCommMembers(epoch metaEpoch) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	delete(m.syncCommsByEpoch, epoch.Epoch)
+}
+
+func (m *Component) slotAttester(slot uint64) *SlotAttester {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	return m.attestersBySlot[slot] // Make nil values valid noops
+}
+
+func (m *Component) syncCommMember(epoch uint64) *SyncCommMember {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	return m.syncCommsByEpoch[epoch]
+}
+
+func (m *Component) isStartup() bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	isStartup := !m.started
+
+	return isStartup
+}
+
+func (m *Component) setAttester(attester *SlotAttester) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.attestersBySlot[uint64(attester.Slot())] = attester
+}
+
+func (m *Component) setSyncCommMember(syncCommMem *SyncCommMember) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.syncCommsByEpoch[uint64(syncCommMem.epoch)] = syncCommMem
+}
+
+// orderByTime sorts the given set of duties by their start times.
+func orderByTime(duties map[scheduleTuple]struct{}) []scheduleTuple {
+	var resp []scheduleTuple
+	for duty := range duties {
+		resp = append(resp, duty)
+	}
+
+	sort.Slice(resp, func(i, j int) bool {
+		return resp[i].startTime.Before(resp[j].startTime)
+	})
+
+	return resp
+}
+
+// offsetFunc that returns the time at which a duty should be triggered for a given slot.
+type offsetFunc func(metaSlot) time.Time
+
+// offsetFuncsByDuty defines the offsets by duty type.
```

Comment:
Offset from what?

I think it's best to document it.

Reply:
Renamed it to a better name.